test/e2e/tests/app_test.go

```diff
@@ -44,15 +44,15 @@ func TestApp_Hash(t *testing.T) {
 		require.NoError(t, err)
 		require.NotEmpty(t, info.Response.LastBlockAppHash, "expected app to return app hash")
 
-		block, err := client.Block(ctx, nil)
-		require.NoError(t, err)
-		require.EqualValues(t, info.Response.LastBlockAppHash, block.Block.AppHash.Bytes(),
-			"app hash does not match last block's app hash")
-
 		status, err := client.Status(ctx)
 		require.NoError(t, err)
 		require.EqualValues(t, info.Response.LastBlockAppHash, status.SyncInfo.LatestAppHash,
 			"app hash does not match node status")
+
```

Comment:
I'm trying to understand how changing the order could affect the outcome here. The Status call doesn't seem to have any access to the data we're comparing. Is the idea that calling Status could change what block we get back from the wrong value to the right one?

What am I missing? 🙂

Reply:
I think this is just an issue of atomicity. Beforehand we just asked for the latest block which could have changed (i.e. a new block may have been produced) in between calling `ABCIInfo` and `Block`. This is solved by specifying the height (being the same as was returned in `ABCIInfo` in the call to retrieve the `Block`. However, the problem still persists. It's conceivable that a new block is committed in between `ABCIInfo` and `Status` in which case we will get different hashes. The best solution I think is that we check the heights match before checking the app hashes.
```
status, err := client.Status(ctx)
require.NoError(t, err)
if info.Response.LastBlockHeight == status.SyncInfo.LastBlockHeight {
    require.EqualValues(t, info.Response.LastBlockAppHash, status.SyncInfo.LatestAppHash,
        "app hash does not match node status")
}
```
There are a few atomicity problems in Tendermint. One example is that we don't batch persisting of blocks and state together. This means occasionally the block store is ahead of the state store and we need to have logic on restarts to handle this. Looking forward to the db work fixing some of this